client_code/persistence.py

```diff
@@ -0,0 +1,166 @@
+#
+# This software is published at https://github.com/anvilistas/anvil-extras
+import anvil.server
+
+__version__ = "2.1.4"
+
+
+def _snakify(text):
+    return "".join("_" + c.lower() if c.isupper() else c for c in text).lstrip("_")
+
+
+class LinkedAttribute:
+    """A descriptor class for adding linked table items as attributes
+
+    For a class backed by a data tables row, this class is used to dyamically add
+    linked table items as attributes to the parent object.
+    """
+
+    def __init__(self, linked_column, linked_attr):
+        """
+        Parameters
+        ----------
+        linked_column: str
+            The name of the column in the row object which links to another table
+        linked_attr: str
+            The name of the column in the linked table which contains the required
+            value
+        """
+        self._linked_column = linked_column
+        self._linked_attr = linked_attr
+
+    def __set_name__(self, owner, name):
+        if name == self._linked_column:
+            raise ValueError(
+                "Attribute name cannot be the same as the linked column name"
+            )
+        self._name = name
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
+
+        if instance._delta:
+            return instance._delta[self._name]
+
+        if not instance._store:
+            return None
+
+        return instance._store[self._linked_column][self._linked_attr]
+
+    def __set__(self, instance, value):
+        instance._delta[self._name] = value
+
+
+class LinkedClass:
+    "A descriptor class for adding objects based on linked tables as attributes"
+
+    def __init__(self, cls, *args, linked_column=None, **kwargs):
+        self._linked_column = linked_column
+        self._cls = cls
+        self._args = args or []
+        self._kwargs = kwargs or {}
+
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
+
+        return self._cls(
+            instance._store[self._linked_column], *self._args, **self._kwargs
+        )
+
+    def __set__(self, instance, value):
+        raise AttributeError(
+            "Linked Class instance is already set and cannot be changed"
+        )
+
+
+class PersistedClass:
+    _cache = {}
+    key = None
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._snake_name = _snakify(cls.__name__)
+        for attr, value in cls.__dict__.items():
+            try:
+                is_persisted_class = issubclass(value, PersistedClass)
+            except TypeError:
+                is_persisted_class = False
+
+            if is_persisted_class:
+                setattr(cls, attr, LinkedClass(cls=value, linked_column=attr))
+
+    @classmethod
+    def search(cls, lazy=False, *args, **kwargs):
+        rows = anvil.server.call(f"search_{cls.__name__.lower()}", *args, **kwargs)
```

Comment:
```suggestion
        rows = anvil.server.call(f"search_{cls._snake_name}", *args, **kwargs)
```

Reply:
Ugh! How did that error creep back in?!!!